corehq/sql_db/util.py

```diff
@@ -150,3 +157,96 @@ def _get_all_nested_subclasses(cls):
         if subclass not in seen:
             seen.add(subclass)
             yield subclass
+
+
+@memoized
+def get_standby_databases():
+    standby_dbs = []
+    for db_alias in settings.DATABASES:
+        with db.connections[db_alias].cursor() as cursor:
+            cursor.execute("SELECT pg_is_in_recovery()")
+            [(is_standby, )] = cursor.fetchall()
+            if is_standby:
+                standby_dbs.append(db_alias)
+    return standby_dbs
+
+
+def get_replication_delay_for_standby(db_alias):
+    """
+    Finds the replication delay for given database by running a SQL query on standby database.
+        See https://www.postgresql.org/message-id/CADKbJJWz9M0swPT3oqe8f9+tfD4-F54uE6Xtkh4nERpVsQnjnw@mail.gmail.com
+
+    If the given database is not a standby database, zero delay is returned
+    If standby process (wal_receiver) is not running on standby a `VERY_LARGE_DELAY` is returned
+    """
+    if db_alias not in get_standby_databases():
+        return 0
+    # used to indicate that the wal_receiver process on standby is not running
+    VERY_LARGE_DELAY = 100000
+    sql = """
+    SELECT
+    CASE
+        WHEN NOT EXISTS (SELECT 1 FROM pg_stat_wal_receiver) THEN {delay}
+        WHEN pg_last_xlog_receive_location() = pg_last_xlog_replay_location() THEN 0
+        ELSE EXTRACT (EPOCH FROM now() - pg_last_xact_replay_timestamp())::INTEGER
+    END
+    AS replication_lag;
+    """.format(delay=VERY_LARGE_DELAY)
+    with db.connections[db_alias].cursor() as cursor:
+        cursor.execute(sql)
+        [(delay, )] = cursor.fetchall()
+        return delay
+
+
+@memoized
+def get_standby_delays_by_db():
+    ret = {}
+    for _db, config in six.iteritems(settings.DATABASES):
+        delay = config.get('HQ_ACCEPTABLE_STANDBY_DELAY')
+        if delay:
+            ret[_db] = delay
+    return ret
+
+
+def filter_out_stale_standbys(dbs):
+    # from given list of databases filters out those with more than
+    #   acceptable standby delay, if that database is a standby
+    delays_by_db = get_standby_delays_by_db()
+    return [
+        db
+        for db in dbs
+        if get_replication_delay_for_standby(db) <= delays_by_db.get(db, ACCEPTABLE_STANDBY_DELAY_SECONDS)
+    ]
+
+
+def select_db_for_read(weighted_dbs):
+    """
+    Returns a randomly selected database per the weights assigned from
+        a list of databases. If any database is standby and its replication has
+        more than accesptable delay, that db is dropped from selection
+
+    Args:
+        weighted_dbs: a list of tuple of db and the weight.
+            [
+                ("pgmain", 5),
+                ("pgmainstandby", 5)
+            ]
+
+    """
+    # convert to a db to weight dictionary
+    weights_by_db = {_db: weight for _db, weight in weighted_dbs}
+
+    # filter out stale standby dbs
+    fresh_dbs = filter_out_stale_standbys(weights_by_db.keys())
```

Comment:
Can we move the caching to this function so we aren't re-computing this list every time.

Reply:
Ha, updated.